net/adguardhome/Makefile

```
@@ -56,6 +57,7 @@ endef
 define Build/Compile
 	( \
 		pushd $(PKG_BUILD_DIR) ; \
+		go mod tidy; \
```

```suggestion
		$(GO_GENERAL_BUILD_CONFIG_VARS) \
		$(GO_PKG_BUILD_CONFIG_VARS) \
		$(GO_PKG_BUILD_VARS) \
		go mod tidy; \
```

If you decide to use this hack, please add Go vars so it won't pollute user's environment at least.

---

Done.

---

Cc: @jefferyto 
I'm not sure which one is the better approach, patch or this kind of hack? I'd prefer the former though.

---

I would prefer a patch as just seeing `go mod tidy` I'm not sure why it is necessary or what it is changing, and it would be very easy for unwanted changes/side effects to creep in in the future.

If the patch would be very large then maybe this can be justified but there would need to be some explanation in the makefile.

---

https://github.com/immortalwrt/packages/blob/c4de559bdaa3f7442a3f8e2deefbd66dcdec42a2/net/adguardhome/patches/010-go-mod-tidy.patch

119 lines, I think it's reasonable.

---

> https://github.com/immortalwrt/packages/blob/c4de559bdaa3f7442a3f8e2deefbd66dcdec42a2/net/adguardhome/patches/010-go-mod-tidy.patch
> 
> 119 lines, I think it's reasonable.

It's still not clear to me, what is the error that this is meant to fix/workaround and what change(s) is necessary for this fix?

---

Go compiler just throws an error:
```
go: updates to go.mod needed; to update it:
        go mod tidy
```

---

I believe downgrading github.com/mdlayher/socket is the correct fix:

```patch
--- a/go.mod
+++ b/go.mod
@@ -48,7 +48,7 @@ require (
 	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/go-task/slim-sprig v0.0.0-20230315185526-52ccab3ef572 // indirect
 	github.com/google/pprof v0.0.0-20230926050212-f7f687d19a98 // indirect
-	github.com/mdlayher/socket v0.5.0 // indirect
+	github.com/mdlayher/socket v0.4.1 // indirect
 	github.com/onsi/ginkgo/v2 v2.13.0 // indirect
 	github.com/patrickmn/go-cache v2.1.0+incompatible // indirect
 	github.com/pierrec/lz4/v4 v4.1.18 // indirect
--- a/go.sum
+++ b/go.sum
@@ -71,8 +71,8 @@ github.com/mdlayher/packet v1.1.2/go.mod
 github.com/mdlayher/raw v0.1.0 h1:K4PFMVy+AFsp0Zdlrts7yNhxc/uXoPVHi9RzRvtZF2Y=
 github.com/mdlayher/raw v0.1.0/go.mod h1:yXnxvs6c0XoF/aK52/H5PjsVHmWBCFfZUfoh/Y5s9Sg=
 github.com/mdlayher/socket v0.2.1/go.mod h1:QLlNPkFR88mRUNQIzRBMfXxwKal8H7u1h3bL1CV+f0E=
-github.com/mdlayher/socket v0.5.0 h1:ilICZmJcQz70vrWVes1MFera4jGiWNocSkykwwoy3XI=
-github.com/mdlayher/socket v0.5.0/go.mod h1:WkcBFfvyG8QENs5+hfQPl1X6Jpd2yeLIYgrGFmJiJxI=
+github.com/mdlayher/socket v0.4.1 h1:eM9y2/jlbs1M615oshPQOHZzj6R6wMT7bX5NPiQvn2U=
+github.com/mdlayher/socket v0.4.1/go.mod h1:cAqeGjoufqdxWkD7DkpyS+wcefOtmu5OQ8KuoJGIReA=
 github.com/miekg/dns v1.1.56 h1:5imZaSeoRNvpM9SzWNhEcP9QliKiz20/dA2QabIGVnE=
 github.com/miekg/dns v1.1.56/go.mod h1:cRm6Oo2C8TY9ZS/TqsSrseAcncm74lfK5G+ikN2SWWY=
 github.com/niemeyer/pretty v0.0.0-20200227124842-a10e7caefd8e h1:fD57ERR4JtEqsWbfPhv4DMiApHyliiK5xCTNVSPiaAs=
```

Go 1.21 is more serious about the `go` version declared in go.mod and has [this rule](https://go.dev/doc/toolchain#config):

> A module’s `go` line must declare a version greater than or equal to the `go` version declared by each of the modules listed in `require` statements...
> 
> For example, if module _M_ requires a dependency _D_ with a `go.mod` that declares `go 1.22.0`, then _M_’s `go.mod` cannot say `go 1.21.3`.

go wants go.mod to be updated because the [`go` line](https://github.com/AdguardTeam/AdGuardHome/blob/v0.107.41/go.mod#L3)  is `go 1.20`, but the [`go` line](https://github.com/mdlayher/socket/blob/v0.5.0/go.mod#L3) for github.com/mdlayher/socket v0.5.0 is `go 1.21`. (v0.5.0 is the first version that [only supports Go 1.21+](https://github.com/mdlayher/socket/blob/18f45b55db258c8db998cab787ca7a11529105b6/CHANGELOG.md#v050).)

I believe upstream made a mistake in updating github.com/mdlayher/socket from v0.4.1 to v0.5.0; I will open a PR with this change for upstream soon.

---

Issue: https://github.com/AdguardTeam/AdGuardHome/issues/6446
PR: https://github.com/AdguardTeam/AdGuardHome/pull/6447